File: core/src/org/rexcellentgames/burningknight/entity/level/save/LevelSave.java
package org.rexcellentgames.burningknight.entity.level.save;

import org.rexcellentgames.burningknight.Dungeon;
import org.rexcellentgames.burningknight.entity.level.Level;
import org.rexcellentgames.burningknight.entity.level.SaveableEntity;
import org.rexcellentgames.burningknight.game.state.LoadState;
import org.rexcellentgames.burningknight.util.file.FileReader;
import org.rexcellentgames.burningknight.util.file.FileWriter;

import java.io.IOException;
import java.util.ArrayList;

public class LevelSave {
	public static ArrayList<SaveableEntity> all = new ArrayList<>();

	public static void remove(SaveableEntity entity) {
		all.remove(entity);
	}

	public static void add(SaveableEntity entity) {
		all.add(entity);
	}

	public static void save(FileWriter writer) {
		try {
			Dungeon.level.save(writer);

			writer.writeInt32(all.size());

			for (SaveableEntity entity : all) {
				writer.writeString(entity.getClass().getName().replace("org.rexcellentgames.burningknight.", ""));
				entity.save(writer);
			}
		} catch (Exception e) {
			e.printStackTrace();
		}
	}

	public static void load(FileReader reader) throws IOException {
		all.clear();
		Dungeon.level = Level.forDepth(Dungeon.depth);
		Dungeon.area.add(Dungeon.level);
		Dungeon.level.load(reader);

		int count = reader.readInt32();

		for (int i = 0; i < count; i++) {
			String t = reader.readString();
			Class<?> clazz = null;

			try {
				clazz = Class.forName("org.rexcellentgames.burningknight." + t);
			} catch (ClassNotFoundException e) {
				e.printStackTrace();
			}

			SaveableEntity entity;

			try {
				entity = (SaveableEntity) clazz.newInstance();
			} catch (InstantiationException e) {
				e.printStackTrace();
				continue;
			} catch (IllegalAccessException e) {
				e.printStackTrace();
				continue;
			}

			Dungeon.area.add(entity);
			all.add(entity);

			entity.load(reader);
		}
	}

	private static int i;

	public static void generate() {
		try {
			LoadState.generating = true;

			Dungeon.level = Level.forDepth(Dungeon.depth);
			Dungeon.area.add(Dungeon.level);
			Dungeon.level.generate(i);

			i = 0;
		} catch (RuntimeException e) {
			e.printStackTrace();
			generate();
			i++;
		}
	}
}

File: BurningKnight/Assets/Saves/LevelSave.cs
using System;
using System.Collections.Generic;
using BurningKnight.Entities;
using BurningKnight.Util.Files;

namespace BurningKnight.Assets.Saves
{
	public static class LevelSave
	{
		private static List<SaveableEntity> list = new List<SaveableEntity>();

		public static void Add(SaveableEntity e)
		{
			list.Add(e);
		}

		public static void Remove(SaveableEntity e)
		{
			list.Remove(e);
		} 
		
		public static void Save(FileWriter stream)
		{
			stream.WriteInt(list.Count);

			for (int i = 0; i < list.Count; i++)
			{
				SaveableEntity e = list[i];

				stream.WriteString();
			}
		}

		public static void Load(FileReader stream)
		{
			list.Clear();
			
			int count = stream.ReadInt();

			if (count == 0)
			{
				stream.Close();
				return;
			}

			for (int i = 0; i < count; i++)
			{
				string id = stream.ReadString();
				Type type = Type.GetType(id);

				if (type == null)
				{
					Log.Error("Failed to load " + id);
					continue;
				}

				SaveableEntity e = (SaveableEntity) type.GetConstructor(null)?.Invoke(null);

				if (e == null)
				{
					Log.Error("Failed to create " + id + " (no empty constructor?)");
					continue;
				}
				
				list.Add(e);
				BurningKnight.Area.Add(e);
				
				e.Load(stream);
			}
			
			stream.Close();
		}

		public static void Generate()
		{
			
		}
	}
}